### DISCUSSION TITLE
Can I somehow pack the all.d.ts into a single file?

### DISCUSSION THREAD

**Discussion by datvm** [material-components/material-web#5129] [Q&A]
I am not familiar with Node and Rollup so please excuse me if this is a simple question. Right now `all.d.ts` is like this:

```ts
import './button/elevated-button.js';
import './button/filled-button.js';
import './button/filled-tonal-button.js';
// ...
export * from './button/elevated-button.js';
export * from './button/filled-button.js';
export * from './button/filled-tonal-button.js';
```

Is there a way to generate a `typings.d.ts` that contains everything in a single file, I would like to use this file in a TypeScript project without NodeJS, maybe like:

```ts
declare global {
    interface HTMLElementTagNameMap {
        'md-elevated-button': MdElevatedButton;
    }

   interface MdElevatedButton {
       // ...
   }

   // ...
}
```

Thank you!

**Comment by asyncliz** [comment]
I don't think it's necessary to bundle typings into a single file in order for a non-nodejs typescript project to use it. Can you share more of your project setup and what isn't working?

**Reply by datvm** [reply]
Hi, most of our projects are not NodeJS but .NET with TypeScript. Therefore we would rather not have a `node_modules` folder in them. We prefer having a single `d.ts` file so we can put it in the project folder instead.

**Reply by asyncliz** [reply]
Hm, I'm [not sure this is possible](https://stackoverflow.com/a/39725322). I don't see any mention of this capability [in the typescript docs](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html) either.

There's a [rollup plugin for `d.ts` files](https://github.com/Swatinem/rollup-plugin-dts) that mentions bundling, you could try that.

**Reply by datvm** [reply]
Ah somehow I failed to Google the first link. I did come across the plugin you mentioned so I guess I need to learn Rollup this weekend and give it a try then. Thanks.

**Reply by datvm** [reply]
So last weekend I tried to bundle them with 3 different Rollup plugins but one didn't work and the other 2 somehow doesn't bundle Lit with them (see https://github.com/wessberg/rollup-plugin-ts/issues/223). I am not sure if I made a mistake somewhere or does special declaration is needed for Lit Components?

**Comment by alexmeier-19** [comment]
@datvm @asyncLiz have you found a solution for this? I'm in the same boat. I bundled all material web into one JS file, and now I can use "import/export" from the @material/web node_module in my custom "material-web.d.ts" to use the types in my project. But then, without "skipLibCheck = true", the compilation fails. We don't want to use "skipLibCkeck" as it brings some disadvantages to our project.
So we want to have on ".d.ts" file with all the declarations "decalre..." so we can use the library without "node_modules" entirely.
I found some tools like "dts-bundle-generator", "npm-dts" and so on, but they all kind of start from building a TS project, which material web isn't is it?
So what can we do? Help is much appreciated

**Reply by datvm** [reply]
@alexmeier-19 Sorry somehow I missed the notification 2 weeks ago. Unfortunately I couldn't solve this problem. If you want typings now the only way is to use NPM `node_modules`.
